File: LearningLinearAlgebra.Tests/Matrices/RealColumnVectorTests.cs
using LearningLinearAlgebra.Matrices.Real;
using LearningLinearAlgebra.Numbers;
using FluentAssertions;
using FluentAssertions.Execution;
using Xunit;

using static LearningLinearAlgebra.Matrices.Real.SquareMatrix<float>;
using static LearningLinearAlgebra.Matrices.Real.ColumnVector<float>;
using static LearningLinearAlgebra.Numbers.RealNumber<float>;

namespace LearningLinearAlgebra.Tests.Matrices;

public class RealColumnVectorTests
{
    [Fact]
    public void Sum_of_two_vectors_is_calculated_as_sum_of_the_components()
    {
        var a = V([-1, -3]);
        var b = V([-7, -13]);

        var sum = Add(a, b);

        using var _ = new AssertionScope();

        sum.Should().Equal(V([-8, -16]));
        (a + b).Should().Equal(Add(a, b));
        a.Add(b).Should().Equal(Add(a, b));
    }

    [Fact]
    public void Sum_of_vectors_is_commutative()
    {
        var a = V([-1, -3]);
        var b = V([-7, -13]);

        (a + b).Should().Equal(b + a);
    }

    [Fact]
    public void Sum_of_vectors_is_associative()
    {
        var a = V([-1, -3]);
        var b = V([-7, -13]);
        var c = V([-23, -31]);

        (a + (b + c)).Should().Equal(a + b + c);
    }

    [Fact]
    public void Sum_of_vector_and_its_the_inverse_is_zero()
    {
        var vector = V([-1, -3]);

        var zero = ColumnVector<float>.Zero(2);

        (vector + -vector).Should().Equal(zero);
    }

    [Fact]
    public void Zero_is_an_additive_identity()
    {
        var vector = V([-1, -3]);
        var zero = ColumnVector<float>.Zero(2);

        using var _ = new AssertionScope();

        (vector + zero).Should().Equal(vector);
        (zero + vector).Should().Equal(vector);
    }

    [Fact]
    public void Difference_of_two_vectors_is_calculated_as_difference_of_the_components()
    {
        var a = V([1, 3]);
        var b = V([7, 13]);

        var difference = Subtract(a, b);

        using var _ = new AssertionScope();

        difference.Should().Equal(V([-6, -10]));
        (a - b).Should().Equal(Subtract(a, b));
    }

    [Fact]
    public void When_multiplying_a_vector_by_scalar_then_each_element_of_the_vector_is_multiplied_by_the_scalar()
    {
        var scalar = R(5);
        var vector = V([11, 19]);

        var multiplied = Multiply(scalar, vector);

        using var _ = new AssertionScope();

        multiplied.Should().Equal(V([55, 95]));
        (scalar * vector).Should().Equal(Multiply(scalar, vector));
    }

    [Fact]
    public void Scalar_multiplication_respects_vector_multiplication()
    {
        var scalarA = R(-3);
        var scalarB = R(-7);
        var vector = V([-23, -31]);

        (scalarA * scalarB * vector).Should().Equal(scalarA * (scalarB * vector));
    }

    [Fact]
    public void Scalar_multiplication_distributes_over_addition()
    {
        var scalar = R(-3);
        var vectorA = V([-7, -13]);
        var vectorB = V([-23, -31]);

        (scalar * vectorA + scalar * vectorB).Should().Equal(scalar * (vectorA + vectorB));
    }

    [Fact]
    public void Scalar_multiplication_distributes_over_vector_addition()
    {
        var scalarA = R(-3);
        var scalarB = R(-7);
        var vector = V([-23, -31]);

        (scalarA * vector + scalarB * vector).Should().Equal((scalarA + scalarB) * vector);
    }

    [Fact]
    public void Transpose_of_a_column_vector_is_row_vector_with_same_entries_of_the_original_vector()
    {
        var vector = V([5, 3, -7]);

        var transpose = ColumnVector<float>.Transpose(vector);

        using var _ = new AssertionScope();

        transpose.Should().Equal(RowVector<float>.V([5, 3, -7]));
        vector.Transpose().Should().Equal(transpose);
    }

    [Fact]
    public void Inner_product_is_a_sum_of_products_of_vector_components()
    {
        var a = V([5, 3, -7]);
        var b = V([6, 2, 0]);

        var innerProduct = InnerProduct(a, b);

        using var _ = new AssertionScope();

        innerProduct.Should().Be(R(36));
        a.InnerProduct(b).Should().Be(InnerProduct(a, b));
        (a * b).Should().Be(InnerProduct(a, b));
    }

    [Fact]
    public void Inner_product_respects_addition()
    {
        var a = V([-1, -3]);
        var b = V([-7, -13]);
        var c = V([-23, -31]);

        (a * c + b * c).Should().Be((a + b) * c);
    }

    [Fact]
    public void Inner_product_respects_scalar_multiplication()
    {
        var a = V([-1, -3]);
        var b = V([-7, -13]);
        var scalar = R(23);

        (scalar * (a * b)).Should().Be(scalar * a * b);
    }

    [Fact]
    public void Norm_is_square_root_of_inner_product_of_vector_with_itself()
    {
        var vector = V([3, -6, 2]);

        var norm = Norm(vector);

        using var _ = new AssertionScope();

        norm.Should().Be(R(7));
        vector.Norm().Should().Be(Norm(vector));
    }

    [Fact]
    public void Vector_can_be_normalized_to_have_length_of_one_by_dividing_it_by_its_length()
    {
        var vector = V([3, -6, 2]);

        var normalized = Normalized(vector);

        using var _ = new AssertionScope();

        normalized.Should().Equal(1 / Sqrt(vector * vector) * vector);
        vector.Normalized().Should().Equal(Normalized(vector));
    }

    [Fact]
    public void Distance_of_the_two_vectors_is_the_norm_of_the_difference()
    {
        var a = V([3, 1, 2]);
        var b = V([2, 2, -1]);

        var distance = Distance(a, b);

        using var _ = new AssertionScope();

        distance.Should().Be(Sqrt(11));
        a.Distance(b).Should().Be(Distance(a, b));
    }

    [Fact]
    public void Tensor_product_of_vectors_contains_combinations_of_products_of_all_elements_of_both_vectors()
    {

        var a = V([3, 4, 7]);
        var b = V([-1, 2]);

        var tensorProduct = TensorProduct(a, b);

        using var _ = new AssertionScope();

        tensorProduct.Should().Equal(V([-3, -4, -7, 6, 8, 14]));
        a.TensorProduct(b).Should().Equal(TensorProduct(a, b));
    }

    [Fact]
    public void Another_example_of_tensor_product()
    {

        var a = V([-1, 2]);
        var b = V([3, 4, 7]);

        var tensorProduct = TensorProduct(a, b);

        tensorProduct.Should().Equal(V([-3, 6, -4, 8, -7, 14]));
    }

    [Fact]
    public void Tensor_product_is_associative()
    {
        var a = V([-1, -3]);
        var b = V([-7, -13]);
        var c = V([-23, -31]);

        TensorProduct(TensorProduct(a, b), c).Should().Equal(TensorProduct(a, TensorProduct(b, c)));
    }

    [Fact]
    public void Vector_is_converted_from_linearly_independent_base_to_orthonormal_base_by_dividing_it_by_its_norm()
    {
        var I = V([3, 0, 0]);
        var II = V([0, 1, 2]);
        var III = V([0, 25]);

        using var _ = new AssertionScope();

        Orthonormal(I).Should().Equal(1 / Sqrt(I * I) * I);
        Orthonormal(II).Should().Equal(1 / Sqrt(II * II) * II);
        Orthonormal(III).Should().Equal(1 / Sqrt(III * III) * III);

        I.Orthonormal().Should().Equal(Orthonormal(I));
    }
}

File: LearningLinearAlgebra.FSharp.Tests/RealColumnVectorTests.fs
namespace LearningLinearAlgebra.Numbers

open Xunit

module RealColumnVectorTests =

    open LearningLinearAlgebra
    open LearningLinearAlgebra.Matrices.Real

    [<Fact>]
    let ``Sum of two vectors is calculated as sum of the components`` () =
        let a = V [| -1.0; -3 |]
        let b = V [| -7.0; -13 |]

        let sum = ColumnVector.Add a b

        Assert.Equal<ColumnVector<float>>(a, a)

        sum |> Should.BeEquivalentTo.Real.ColumnVector(V [| -8.0; -16 |])
        a + b |> Should.BeEquivalentTo.Real.ColumnVector(ColumnVector.Add a b)

    [<Fact>]
    let ``Sum of vectors is commutative`` () =
        let a = V [| -1.0; -3 |]
        let b = V [| -7.0; -13 |]

        b + a |> Should.BeEquivalentTo.Real.ColumnVector(a + b)

    [<Fact>]
    let ``Sum of vectors is associative`` () =
        let a = V [| -1.0; -3 |]
        let b = V [| -7.0; -13 |]
        let c = V [| -23.0; -31 |]

        a + (b + c) |> Should.BeEquivalentTo.Real.ColumnVector((a + b) + c)

    [<Fact>]
    let ``Sum of vector and it's the inverse is zero`` () =
        let vector = V [| -1.0; -3 |]

        let zero = ColumnVector.Zero 2

        vector + (-vector) |> Should.BeEquivalentTo.Real.ColumnVector(zero)

    [<Fact>]
    let ``Zero is an additive identity`` () =
        let vector = V [| -1.0; -3 |]
        let zero = ColumnVector.Zero 2

        vector + zero |> Should.BeEquivalentTo.Real.ColumnVector(vector)
        zero + vector |> Should.BeEquivalentTo.Real.ColumnVector(vector)

    [<Fact>]
    let ``Difference of two vectors is calculated as difference of the components`` () =
        let a = V [| 1.0; 3 |]
        let b = V [| 7.0; 13 |]

        let difference = ColumnVector.Subtract a b

        difference |> Should.BeEquivalentTo.Real.ColumnVector(V [| -6.0; -10 |])
        a - b |> Should.BeEquivalentTo.Real.ColumnVector(ColumnVector.Subtract a b)

    [<Fact>]
    let ``When multiplying a vector by scalar then each element of the vector is multiplied by the scalar`` () =
        let scalar = 5.0
        let vector = V [| 11.0; 19.0 |]

        let multiplied = ColumnVector.Multiply scalar vector

        multiplied |> Should.BeEquivalentTo.Real.ColumnVector(V [| 55.0; 95.0 |])

        scalar * vector
        |> Should.BeEquivalentTo.Real.ColumnVector(ColumnVector.Multiply scalar vector)

    [<Fact>]
    let ``Scalar multiplication respects vector multiplication`` () =
        let scalarA = -3.0
        let scalarB = -7.0
        let vector = V [| -23.0; -31.0 |]

        (scalarA * scalarB) * vector
        |> Should.BeEquivalentTo.Real.ColumnVector(scalarA * (scalarB * vector))

    [<Fact>]
    let ``Scalar multiplication distributes over addition`` () =
        let scalar = -3.0
        let vectorA = V [| -7.0; -13.0 |]
        let vectorB = V [| -23.0; -31.0 |]

        (scalar * vectorA) + (scalar * vectorB)
        |> Should.BeEquivalentTo.Real.ColumnVector(scalar * (vectorA + vectorB))

    [<Fact>]
    let ``Scalar multiplication distributes over vector addition`` () =
        let scalarA = -3.0
        let scalarB = -7.0
        let vector = V [| -23.0; -31.0 |]

        (scalarA * vector) + (scalarB * vector)
        |> Should.BeEquivalentTo.Real.ColumnVector((scalarA + scalarB) * vector)

    [<Fact>]
    let ``Inner product is a sum of products of vector components`` () =
        let a = V [| 5.0; 3.0; -7.0 |]
        let b = V [| 6.0; 2.0; 0.0 |]

        let innerProduct = ColumnVector.InnerProduct a b

        innerProduct |> Should.BeEquivalentTo.Real.Number(36.0)
        a * b |> Should.BeEquivalentTo.Real.Number(ColumnVector.InnerProduct a b)

    [<Fact>]
    let ``Inner product respects addition`` () =
        let a = V [| -1.0; -3.0 |]
        let b = V [| -7.0; -13.0 |]
        let c = V [| -23.0; -31.0 |]

        (a * c) + (b * c) |> Should.BeEquivalentTo.Real.Number((a + b) * c)

    [<Fact>]
    let ``Inner product respects scalar multiplication`` () =
        let a = V [| -1.0; -3.0 |]
        let b = V [| -7.0; -13.0 |]
        let scalar = 23.0

        scalar * (a * b) |> Should.BeEquivalentTo.Real.Number((scalar * a) * b)

    [<Fact>]
    let ``Norm is square root of inner product of vector with itself`` () =
        let vector = V [| 3.0; -6.0; 2 |]

        let norm = ColumnVector.Norm vector

        norm |> Should.BeEquivalentTo.Real.Number(7.0)

    [<Fact>]
    let ``Distance of the two vectors is the norm of the difference`` () =
        let a = V [| 3.0; 1.0; 2 |]
        let b = V [| 2.0; 2.0; -1 |]

        let distance = ColumnVector.Distance a b

        distance |> Should.BeEquivalentTo.Real.Number(sqrt (11.0))

    [<Fact>]
    let ``Tensor product of vectors contains combinations scalar products of all elements of both vectors`` () =

        let a = V [| 3.0; 4.0; 7 |]
        let b = V [| -1.0; 2 |]

        let tensorProduct = ColumnVector.TensorProduct a b

        tensorProduct
        |> Should.BeEquivalentTo.Real.ColumnVector(V [| -3.0; 6.0; -4.0; 8.0; -7.0; 14 |])

    [<Fact>]
    let ``Another example of tensor product`` () =

        let a = V [| -1.0; 2 |]
        let b = V [| 3.0; 4.0; 7 |]

        let tensorProduct = ColumnVector.TensorProduct a b

        tensorProduct
        |> Should.BeEquivalentTo.Real.ColumnVector(V [| -3.0; -4.0; -7.0; 6.0; 8.0; 14.0 |])

    [<Fact>]
    let ``Tensor product is associative`` () =
        let a = V [| -1.0; -3 |]
        let b = V [| -7.0; -13 |]
        let c = V [| -23.0; -31 |]

        ColumnVector.TensorProduct a (ColumnVector.TensorProduct b c)
        |> Should.BeEquivalentTo.Real.ColumnVector(ColumnVector.TensorProduct (ColumnVector.TensorProduct a b) c)

    [<Fact>]
    let ``Convert from linearly independenr base to orthonormal base`` () =
        let I = V [| 3.0; 0.0; 0.0 |]
        let II = V [| 0.0; 1.0; 2.0 |]
        let III = V [| 0.0; 25.0 |]

        let norm = (ColumnVector.Norm I)
        let a = (1.0 / norm) * I
        V [| 1.0; 0.0; 0.0 |] |> Should.BeEquivalentTo.Real.ColumnVector(a)
